coalahtml/tree.py

```diff
@@ -0,0 +1,84 @@
+from collections import OrderedDict
+from queue import Queue
+import os
+
+
+class Node:
+    """
+    Node class
+    """
+    count = 1
+
+    def __init__(self, path):
+        self.child = {}
+        self.parent = None
+        self.path = path
+        self.abspath = ""
+        self._id = Node.count
+        Node.count += 1
+
+
+class Tree:
+
+    def __init__(self, file_list):
+        """
+        Create a root and build the graph.
+        """
+        self.root = Node(os.path.commonprefix(file_list))
+        self.root.abspath = self.root.path
+        self.fl_split = Tree._split_paths(file_list)
+        self._build(self.fl_split)
+
+    @staticmethod
+    def _split_paths(path_list):
+        return [path.split('/') for path in path_list]
+
+    def _add_node(self, new_node, prev_node, head):
+        q = Queue()
+        q.put(head)
+
+        while not q.empty():    # Search for parent node: BFS
+            current = q.get()
+            if current._id != prev_node._id:
+                for child in current.child.values():
+                    q.put(child)
+            else:
+                prev_node = current
```

Comment:
So while I am searching two things happen eventually:
* I find the node I am looking for. I insert new node to this current node. But I assign current node to parent node to maintain uniformity in next child insertion step.
* I don't find - In this case I am inserting new node to `prev_node` as it's parent in this case. 

I could restructure the code and make it like
```python
current = None   # Newly added
found_node = False # Newly added
while not q.empty():    # Search for parent node: BFS
            current = q.get()
            if current._id != prev_node._id:
                for child in current.child.values():
                    q.put(child)
            else:
                found_node = True
                break
if found_node is None:
       if prev_node.child.get(new_node.path, None) is None:
            prev_node.child[new_node.path] = new_node
       else:
            # Duplicates
            Node.count -= 1
            return prev_node.child[new_node.path]
else: # Newly added else block.
       if current.child.get(new_node.path, None) is None:
            current.child[new_node.path] = new_node
       else:
            # Duplicates
            Node.count -= 1
            return current.child[new_node.path]
return new_node

```

I have reused the `prev_node` just to avoid creating this unnecessarily verbose code. I will change the style it if you still suggest? :smile: 
@Uran198